uniforce.lua

```diff
@@ -0,0 +1,259 @@
+-- Prompt units to adjust their uniform.
+local help = [====[
+
+uniforce
+========
+
+Prompt units to reevaluate their uniform, by removing/dropping potentially conflicting worn items.
+
+Unlike a "replace clothing" designation, it won't remove additional clothing
+if it's coexisting with a uniform item already on that body part.
+It also won't remove clothing (e.g. shoes, trousers) if the unit has yet to claim an
+armor item for that bodypart. (e.g. if you're still manufacturing them.)
+
+By default it simply prints info about the currently selected unit,
+to actually drop items, you need to provide it the -drop option.
+
+The default algorithm assumes that there's only one armor item assigned per body part,
+which means that it may miss cases where one piece of armor is blocked but the other
+is present. The -multi option can possibly get around this, but at the cost of ignoring
+left/right distinctions when dropping items.
+
+In some cases, an assigned armor item can't be put on because someone else is wearing/holding it.
+The -free option will cause the assigned item to be removed from the container/dwarven inventory
+and placed onto the ground, ready for pickup.
+
+In no cases should the command cause a uniform item that is being properly worn to be removed/dropped.
+
+Targets:
+:(no target): Force the selected dwarf to put on their uniform.
+:-all:        Force the uniform on all military dwarves.
+
+Options:
+:(none):      Simply show identified issues (dry-run).
+:-drop:       Cause offending worn items to be placed on ground under unit.
+:-free:       Remove to-equip items from containers or other's inventories, and place on ground.
+:-multi:      Be more agressive in removing items, best for when uniforms have muliple items per body part.
+]====]
+
+local utils = require('utils')
+
+local validArgs = utils.invert({
+  'all',
+  'drop',
+  'free',
+  'multi',
+  'help'
+})
+
+-- Functions
+
+function get_item_pos(item)
+  local x, y, z = dfhack.items.getPosition(item)
+  if x == nil or y == nil or z == nil then
+    return nil
+  end
+
+  if not dfhack.maps.isValidTilePos(x,y,z) then
+    dfhack.println("NOT VALID TILE")
+    return nil
+  end
+  if not dfhack.maps.isTileVisible(x,y,z) then
+    dfhack.println("NOT VISIBLE TILE")
+    return nil
+  end
+  return {x=x, y=y, z=z}
+end
+
+function find_item_by_ID(id)
+  for k,v in pairs(df.global.world.items.all) do
+    if v.id == id then
+      return v
+    end
+  end
+  return nil
+end
+
+function find_squad_position(unit)
+  for i, squad in pairs( df.global.world.squads.all ) do
+    for i, position in pairs( squad.positions ) do
+      if position.occupant == unit.hist_figure_id then
+        return position
+      end
+    end
+  end
+  return nil
+end
+
+-- This should be okay for dwarves, not sure if valid for others.
```

Comment:
Probably unlikely that it'll work for _all_ other races, although humans might be similar enough. It would be great if this could be looked up somehow, but I'm no expert on this. It appears that the caste raws have quite a bit of detail about body parts/layers/etc, accessible through e.g. `df.creature_raw.find(unit.race).caste[unit.caste].body_info` - maybe some of that is relevant. For instance, `body_info.body_parts[15]` of an unmodded dwarf has `category = "FOOT"` and `flags.STANCE = true` (as well as two unidentified flags set, 28 and 29).